engine/classes/Elgg/EntityIconService.php

```diff
@@ -185,81 +185,132 @@ public function saveIconFromElggFile(ElggEntity $entity, ElggFile $file, $type =
-					$entity->x2 = $x2;
-					$entity->y2 = $y2;
-				}
-			}
-			$this->hooks->trigger("entity:$type:saved", $entity->getType(), [
-				'entity' => $entity,
-				'x1' => $x1,
-				'y1' => $y1,
-				'x2' => $x2,
-				'y2' => $y2,
-			]);
-			return true;
-		};
-
-		$fail = function() use ($entity, $type) {
-			$this->deleteIcon($entity, $type);
-			return false;
-		};
-
+		
+		$x1 = (int) elgg_extract('x1', $coords);
+		$y1 = (int) elgg_extract('y1', $coords);
+		$x2 = (int) elgg_extract('x2', $coords);
+		$y2 = (int) elgg_extract('y2', $coords);
+		
 		$created = $this->hooks->trigger("entity:$type:save", $entity_type, [
 			'entity' => $entity,
 			'file' => $file,
 			'x1' => $x1,
 			'y1' => $y1,
 			'x2' => $x2,
 			'y2' => $y2,
-				], false);
+		], false);
 
-		if ($created === true) {
-			return $success();
+		// did someone else handle saving the icon?
+		if ($created !== true) {
+			// remove existing icons
+			$this->deleteIcon($entity, $type, true);
+			
+			// save master image
+			$store = $this->generateIcon($entity, $file, $type, $coords, 'master');
+			
+			if (!$store) {
+				$this->deleteIcon($entity, $type);
+				return false;
+			}
 		}
 
-		$sizes = $this->getSizes($entity_type, $entity_subtype, $type);
-
+		if ($type == 'icon') {
+			$entity->icontime = time();
+			if ($x1 || $y1 || $x2 || $y2) {
+				$entity->x1 = $x1;
+				$entity->y1 = $y1;
+				$entity->x2 = $x2;
+				$entity->y2 = $y2;
+			}
+		} else {
+			if ($x1 || $y1 || $x2 || $y2) {
+				$entity->{"{$type}_coords"} = serialize([
+					'x1' => $x1,
+					'y1' => $y1,
+					'x2' => $x2,
+					'y2' => $y2,
+				]);
+			}
+		}
+		
+		$this->hooks->trigger("entity:$type:saved", $entity->getType(), [
+			'entity' => $entity,
+			'x1' => $x1,
+			'y1' => $y1,
+			'x2' => $x2,
+			'y2' => $y2,
+		]);
+		
+		return true;
+	}
+	
+	/**
+	 * Generate an icon for the given entity
+	 *
+	 * @param ElggEntity $entity    Temporary ElggFile instance
+	 * @param ElggFile   $file      Temporary ElggFile instance
+	 * @param string     $type      The name of the icon. e.g., 'icon', 'cover_photo'
+	 * @param array      $coords    An array of cropping coordinates x1, y1, x2, y2
+	 * @param string     $icon_size The icon size to generate (leave empty to generate all supported sizes)
+	 *
+	 * @return bool
+	 */
+	protected function generateIcon(ElggEntity $entity, ElggFile $file, $type = 'icon', $coords = [], $icon_size = '') {
+		
+		if (!$file->exists()) {
+			$this->logger->error('Trying to generate an icon from a non-existing file');
+			return false;
+		}
+		
+		$x1 = (int) elgg_extract('x1', $coords);
+		$y1 = (int) elgg_extract('y1', $coords);
+		$x2 = (int) elgg_extract('x2', $coords);
+		$y2 = (int) elgg_extract('y2', $coords);
+		
+		$cropping_mode = ($x2 > $x1) && ($y2 > $y1);
+		
+		$sizes = $this->getSizes($entity->getType(), $entity->getSubtype(), $type);
 		foreach ($sizes as $size => $opts) {
+			
+			if (!empty($icon_size) && ($icon_size !== $size)) {
```

Comment:
generate icon will report true if requested size is not in sizes.. is this correct?

Reply:
You don't need to loop, just extract `$icon_size` from `$sizes` array and bail if not set. Note that `original` might be an empty array.

Reply:
if `$icon_size` is empty all icons will be generated. Also i added a check that if `$icon_size` is given it has to exist in the allowed sizes.